[STEP 2] 은행창구 관리앱 ― Miro, Jenna

ueunli opened the pull request:
> 안녕하세요 제이크(@jryoun1)~! 🙌🏻 
> 저번 Step1 리뷰를 통해 너무 좋은 피드백들을 받고, 많이 배워가서 너무 좋았어요\~ 감사해요 !! 
이번 Step2도 아직 아쉬운 부분들이 있지만, 요구사항을 읽어보고 각 객체의 역할과 책임에 대해서 고민을 많이 해봤어요! 감사합니다! 🙇🏻‍♂️🙇🏻‍♀️

　
# 구현사항
은행원이 하나인 은행을 관리하기 위한 타입 구현 & 콘솔앱 구현
- Queue의 활용
- 타입 추상화 및 일반화

### 👾 타입 구현
> 요구사항: 은행, 고객 타입 및 그 외 필요하다고 생각되는 타입 구현
- 은행
   - [x] 은행이 문을 열면 10~30명의 고객이 Queue에 대기하며, 
     은행은 고객을 선입(등록순번대로), 선출(은행원에게 보내기)함
   - [x] 은행이 문을 닫는 시점은 모든 고객의 업무가 끝났을 때이며, 
     마감 시 당일 고객 수와 총 영업시간을 알리는 안내문구를 출력함
- 은행원
   - [x] 은행원은 고객을 1:1로만 응대하며, 한 고객의 업무를 처리하는 데에는 0.7초가 소요됨
   - [x] 은행원은 고객 응대업무의 시작과 끝에 진행상황을 보고(시작·완료문구 출력)

### 👾 콘솔앱 구현
> 요구사항: 메뉴를 선택하여 은행을 열고 닫거나 프로그램을 종료하는 시스템 구현
- 앱 실행 시
   - 앱을 실행하면 두 개의 메뉴를 출력
      ```Swift
      1 : 은행 개점
      2 : 종료
      ```
   - 메뉴 출력 직후 사용자 입력을 받음
      ```Swift
      입력: `readLine()`
      ```
- 입력 시
   - 사용자가 `1`을 입력하면: 은행 개점 ~ 폐점
      - 은행을 개점하고 10명~30명 사이 임의의 인원이 방문
      - 모든 고객의 업무를 처리하면 은행문이 닫히고 다시 메뉴를 출력
   - 사용자가 `2`를 입력하면: 은행 앱 종료

　
# 고민과 해결
### 🌱 은행원
> Bank의 은행원 프로퍼티를 요소가 하나인 BankClerk배열로 선언하였고, 이유는 아래와 같아요 ↓
- 현재 스텝의 요구사항 명세는 '**은행원을 n명 갖는 은행 타입을 구현**하되 **은행원이 1명뿐인 은행 구현체로 실행**할 것' 이에요
- Bank가 `BankClerk`을 갖기에는 전자에 맞지 않고, 요소가 하나인 `[BankClerk]`을 가지면 조건에는 맞겠지만 적합한 방식인지는 확신이 서지 않았어요. 따라서 유연한 코드를 위해서는 `BankClerk`타입 자체의 역할을 분명히 할 필요가 있다고 생각되어 고민이 많았어요
- 하지만 다음 스텝에서 은행원 또는 업무의 종류가 늘어난다면 그에 맞춰 본격적으로 논의하는 편이 좋겠다고 판단하였고, 일단 `[BankClerk]`타입으로 구현하여 서브스크립트(`[0]`)로 접근하도록 해 두었어요
- 또한, 비동기 처리에 대해서 이후 스텝에서 더 깊게 고민해보고 해당 로직을 리팩토링할 예정이에요.

### 🌱 `usleep()`
- Step2의 요구사항에서는 한 명의 은행원만 존재하기에 따로 비동기 메소드를 사용하지 않고, `usleep()`을 활용해주었어요. 여러 명의 은행원들이 존재하는 Step3에서는 비동기 처리를 해줄 예정이에요!
- 현재 `usleep()`의 파라미터로 들어가는 `timeSpent` 상수는 따로 하드코딩된 값을 모아둔 `BankManager` 구조체 말고 같은 `BankClerk` 구조체 안에 프로퍼티 값으로 설정해 주었어요. 이와 같이 비교적 간단한 하드 코딩 값을 같은 구조체 속 프로퍼티 값으로 빼주는 것이 좋을까요, 아니면 `BankManager`  구조체와 같이 프로젝트 내 하드코딩된 값을 모아두는 곳에 두는 것이 좋을까요?🤔

### 🌱 `BankManager`의 역할
현재 저희 코드에서 `BankManager` 객체의 역할은 아래와 같아요.
- 사용자가 console 창에서 보는 메뉴/종료 출력문
- `Bank` 클래스의 `numberOfCustomers` 프로퍼티에 할당할 랜덤한 손님의 숫자

저희가 이번 스텝을 진행하면서 분리가 필요하지만 어디에 어떻게 모아두어야 깔끔할지 갈피를 잡지 못한 3가지가 있었어요: `Bank가 수행하지 않아도 될 것들`, `하드코딩된 매직스트링이나 매직넘버들`, `콘솔앱 자체의 실행 및 종료를 위한 로직들`
프로젝트 초안에 포함되어 있었던 BankManager라는 파일에 (Bank와 차별화된)어떤 역할의 어떤 타입을 구현할지 고민이 많았는데, 우선 위 3가지를 은행(앱)의 영업지원용 리소스라는 틀에 억지로 묶어 BankManager구조체로 구현하게 되었어요.
제이크가 BankManager라는 타입을 구현한다면 어떤 역할을 맡기고 싶은지, 저희의 BankManager 로직을 어떻게 나누면 좋을 것 같은지 조언 부탁드려요! (enum 및 연산타입프로퍼티로 구현한 부분도 제이크가 보시기에 어떤지 궁금해요🤔)

　
　

longlivedrgn committed:
feat: BankClerk 구조체 생성

ueunli committed:
feat: Bank 구조체 생성

- queue, clerk, numberOfCustomers 프로퍼티 생성
- receive(numberOfCustomers:), lineUpCustomersInQueue(), distributeCustomersToClerk(), handleAllCustomers() 메서드 구현

longlivedrgn committed:
feat: main 파일 로직 구현

- 사용자의 입력값에 따라서 콘솔 앱을 실행 또는 종료시키는 로직 구현

ueunli committed:
chore: 불필요한 키워드 생략

- Bank구조체 일부 메서드의 mutating 키워드 삭제

longlivedrgn committed:
feat: BankManager 구현

- 자주 쓰이는 안내문구 또는 값을 열거형으로 묶어 정의

ueunli committed:
Merge branch 'step2' of https://github.com/ueunli/ios-bank-manager into step2

ueunli committed:
chore: 가독성 개선

- Bank의 이니셜라이저 추가
- 업무 소요시간 수정

longlivedrgn committed:
feat: 접근 제어자 추가

ueunli committed:
chore: 고객 인원수 범위 수정

longlivedrgn committed:
refactor: Bank 클래스 속 clerk  프로퍼티 변경

- 기존의 clerk 상수를 배열 형태로 여러 개의 BankClerk 객체를 담을 수 있는 clerks 상수로 변경

ueunli committed:
chore: 옵셔널 언래핑 방법 수정

jryoun1 commented on BankManager.swift:
```diff
@@ -5,3 +5,38 @@
 
 import Foundation
+
+struct BankManager {
```
BankManager의 역할에 대해서 고민을 많이 해보셨네요 👏

두 분이서 고민하신 것도 좋은 것 같네요. 
다만 저라면 사용자에게 입력을 받거나, 은행에 갈 손님의 수를 정하는 것은 은행매니저라는 타입명과는 어울리지는 않는다고 생각합니다. 뭔가 콘솔앱에서 실행을 위한 정보들을 생성하고 넘겨주는 느낌이다보니까 오히려 콘솔매니저라고 볼 것 같네요. 

BankManager라는 것에 어떠한 역할을 맡길지는 두 분이서 좀 더 고민해보시면 좋을 것 같고, 이때 BankManager가 꼭 필요한 타입일까도 고려해보시면 좋을 것 같아요 ㅎㅎ (필요없다 이런 의미는 아닙니다!! ❌ )

ueunli replied:
Bank를 구현하다 보니 Bank 외부에서 처리해줘야 알맞을 듯한 작업이 몇 가지 나왔어요. 말씀하신 것처럼 콘솔앱의 실행을 위한 데이터·기능, 개점 이전의 은행이 알 수 없는 정보 등이었는데요, 

제시된 빈 BankManager파일의 존재를 무의식적으로 ‘BankManager라는 타입을 만들라’ 라는 의미로서 받아들였던 것 같아요. ‘BankManager’라는 타입만의 역할부터 충분히 고민하지 못하고 역으로 생각하는(남은 역할을 떠안게 됨) 과정에서 많은 부분을 놓치게 되었네요 🥲

수행하는 역할을 고려한 결과, 대부분이 콘솔 관련이라는 결론이 나왔어요. 

- 이에 기존의 BankManager 네이밍을 ConsoleManager로 수정하고, (해당커밋: 347f0b7)
- ‘폐점 메시지 출력’기능은 Bank의 메서드로, (해당커밋: 78add46)
- 은행에 갈 ‘최소·최대·무작위 인원’에 대한 데이터와 기능은 Bank의 프로퍼티와 생성자에 맡겼어요. (해당커밋: 1ddcc96)

위와 같이 수정함로써 객체지향적으로 보완해 보았어요!

jryoun1 commented on BankManager.swift:
```diff
@@ -5,3 +5,38 @@
+        case close = "2"
+        case ask = "입력: "
+        
+        var description: String {
+            switch self {
+            case .open: return self.rawValue + ": 은행개점"
+            case .close: return self.rawValue + ": 종료"
+            case .ask: return self.rawValue
+            }
```
깔끔하고 좋네요 👍🏻

jryoun1 commented on BankManager.swift:
```diff
@@ -5,3 +5,38 @@
+        print("\(Menu.ask.description)", terminator: "")
+    }
+    
+    static func closingMessage(totalNumberOfCustomers: Int, totalTime: Double) {
```
사용하는 부분에서 아래와 같이 사용되고 있는데, 매개변수명을 좀 더 개선해보면 좋을 것 같습니다. 🙏🏻
```swift
BankManager.closingMessage(totalNumberOfCustomers: numberOfCustomers, totalTime: totalTime)

// 예를 들면 아래와 같이 뭔가 함수명을 읽었을 때 영어 문장을 읽듯이 읽힌다면 좀 더 가독성이 좋아질 것 같네요 
BankManager.printClosingMessage(about: totalCustomers, with: totalCosumeTime)  
```

longlivedrgn replied:
조언해 주신 대로 argument label을 활용하면 더욱 가독성 있는 호출구문이 될 것 같아 아래와 같이 수정해 봤어요 !

- 이전: `closingMessage(totalNumberOfCustomers: Int, totalTime: Double)`
- 수정: `printClosingMessage(about totalNumberOfCustomers: Int, with totalConsumedTime: Double)`
> 해당 커밋: 50c273e

jryoun1 commented on BankManagerConsoleApp/BankManagerConsoleApp.xcodeproj/project.pbxproj:
```diff
@@ -88,11 +96,13 @@
 			isa = PBXGroup;
 			children = (
-				C7D65D1A259C8190005510E0 /* BankManager.swift */,
 				C7694E79259C3EC00053667F /* main.swift */,
+				C7D65D1A259C8190005510E0 /* BankManager.swift */,
 				87BAAE6229A455A9006D578A /* Node.swift */,
 				87BAAE6429A455CC006D578A /* SinglyLinkedList.swift */,
 				3BC82CAA29A45E5A00D0B86A /* Queue.swift */,
+				876A7DF529A7335700D26878 /* BankClerk.swift */,
+				3BC82CBE29A735D400D0B86A /* Bank.swift */,
```
파일들을 그룹핑하는 기준이 있을까요? 🙋‍♂️

longlivedrgn replied:
따로 그룹핑을 진행하지 않았었지만, 제이크의 코멘트를 듣고 그룹핑을 한번 해봤어요! 먼저 프로젝트 속 모델에 사용되는 기본 데이터 구조인 `Node`, `SinglyLinkedList`, `Queue`를 **Data Structure** 폴더로 묶어보았어요! 그리고, Bank와 관련된 `BankClerk`과 `Bank` 파일은 **Bank** 폴더로 묶어주고, 나머지 console창에서 사용되는 파일인 `ConsoleManager`는 **Console** 파일로 묶어줬어요. 프로젝트의 실행부인 `main` 파일은 어느 폴더에도 속하지 않는다고 판단이 들어 따로 그룹핑을 해주지 않았어요! 제이크가 보시기에 이와 같은 그룹핑은 어떤 지 궁금해요!
> 해당 커밋: 16e1681

jryoun1 replied:
넵 괜찮아보입니다! 👍

jryoun1 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift:
```diff
@@ -0,0 +1,49 @@
+    private var numberOfCustomers: Int = 0
+    
+    init() {
+        self.queue = Queue<String>()
+        self.clerks = [BankClerk()]
```
Queue의 경우에는 Bank가 가지고 있다고 보면 내부에서 init되어도 될 것 같은데, 
은행원의 경우에는 외부에서 주입을 받지 않을 이유가 있을까요? 
```swift
// 여기서 주입이라는 것은 init()의 매개변수로서 clerks: [BankClerk]을 받는 것을 의미합니다 🙏🏻
init(clerks: [BankClerk]) {
   self.clerks = clerks
}
```

ueunli replied:
와 Jake의 말을 들어보니 정말 은행원을 외부에서 주입받는 편이 타당한 것 같아요 !!
Bank생성 시 외부에서 원하는 은행원을 주입할 수 있게끔 이니셜라이저가 매개변수로 [BankClerk]을 받도록 수정해 주었어요 😄 

> 해당 커밋: 2175f87

jryoun1 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift:
```diff
@@ -0,0 +1,49 @@
+    }
+    
+    mutating func receive(numberOfCustomers: Int) {
+        self.numberOfCustomers = numberOfCustomers
```
numberOfCustomers도 초기화시점이 아닌 따로 함수를 두신 이유가 궁금합니다 🙋‍♂️

ueunli replied:
외부에서 랜덤으로 손님 숫자를 생성하여 `Bank` 객체에게 넣어주려고했어요! 
그러나, `Bank`가 본인이 감당할 수 있는 최대 손님 숫자, 최소 손님 숫자 정보를 가지고 있는 게 더 알맞은 접근이라는 생각이 들어, `Bank` 안에 튜플 형태의 rangeOfNumberOfCustomers 상수를 생성하여 최대 손님 숫자, 최소 손님 숫자 정보를 가질 수 있게 했어요. 또한, `Bank` 객체가 생성이될 때 random 메소드를 활용하여 랜덤한 손님 숫자를 `Bank` 객체가 가질 수 있도록 코드를 수정했어요!

> 해당 커밋: 1ddcc96

jryoun1 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift:
```diff
@@ -0,0 +1,49 @@
+import Foundation
+
+struct Bank {
+    private let queue: Queue<String>
+    private let clerks: [BankClerk]
```
함수명, 접근제어로 좋네요 👏💯

jryoun1 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank.swift:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+    
+    private func distributeCustomersToClerk() -> Node<String>? {
+        let node = queue.dequeue()
+        return node
+    }
```
함수명을 보면 함수 내부에서 clerk에게 customer를 배정해 줄 것 같은데, 
실제로 내부에서는 손님을 꺼내오는 것만 하고 있는 것 같습니다! 
함수명의 변경이나 내부 로직의 변경이 필요해보입니다 🙏🏻

longlivedrgn replied:
현재 `distributeCustomersToClerk()`의 메소드 명에는 분배한다는 의미도 포함되어있기에 기존의 메소드 명을 `extractCustomerFromQueue()`로 변경함으로써 수행 로직을 잘 나타내도록 + 동일계층의 유사기능 메서드와 짝이 맞게끔 수정을 해보았어요!
> 해당 커밋: bdc8fd7

jryoun1 commented on BankManagerConsoleApp/BankManagerConsoleApp/BankClerk.swift:
```diff
@@ -0,0 +1,18 @@
+struct BankClerk {
+    private let timeSpent: useconds_t = 700000
+    
+    func serve(customer: Node<String>) {
```
이때도 매개변수명을 wildcard를 사용하면 좀 더 가독성이 좋아질 것 같습니다.

ueunli replied:
함수 호출 부에서 매개변수명과 인수이름(지역변수)이 일치하여 표현이 불필요하게 중복되므로, wildcard를 적용하여 가독성을 높였습니다!

> 해당 커밋: bdc8fd7

jryoun1 commented on BankManagerConsoleApp/BankManagerConsoleApp/main.swift:
```diff
@@ -5,3 +5,20 @@
+while (chosenMenu ?? "") != BankManager.Menu.close.rawValue {
+    let numberOfCustomers = BankManager.ExpectedNumberOfCustomers.random
+    
+    bank.receive(numberOfCustomers: numberOfCustomers)
+    bank.lineUpCustomersInQueue()
+    bank.handleAllCustomers()
```
어떤 기준으로 bank 내부의 함수들의 접근제어를 주신지 궁금합니다 🙋‍♂️

longlivedrgn replied:
Bank 외부에서 호출되는 Bank 내 메소드는 따로 접근제어를 설정해주지 않았고(internal), 내부에서만 사용되는 메소드는 private 설정을 해주었습니다!😆

jryoun1 replied:
해당 질문을 했던 이유는 과연 외부에서 lineUpCustomersInQueue(), handleAllCustomers()와 같은 동작들에 대해서 알아야하는지에 대한 질문이었습니다. 만약 위와 같은 함수들이 외부에서 알아야할 필요가 있다면 어떤 이유일까요? 

외부에서는 단지 은행이 열렸다 / 닫혔다 정도 혹은 은행이 업무를 시작했다 / 끝냈다 정도만 알아도 될 것 같아보이는데 어떻게 생각하시나요? 

예를 들어 어떤 세탁기라는 개체가 있는데, 위의 함수들의 경우에는 `유저가 넣은 세제를 세탁할 옷들과 물과 함께 섞는다`와 같이 구체적인 동작을 나타내는 것과 같아보이고, 과연 이러한 구체적인 동작을 외부에서 알 필요가 있을까요? 
실제 해당 세탁기라는 개체를 사용하는 입장에서는 세탁기의 세탁 시작 버튼을 누른다와 같이 구체적인 동작보다는 해당 동작들을 추상화시켜 정확히 내부에서는 어떠한 일들이 일어나는지 알 필요가 없다면 알리지 않는 것이 좋아보입니다.

jryoun1 reviewed (changes requested):
Step2 고생하셨습니다~

두 분이서 많은 고민을 해보시고 코드를 작성하시는 것 같아 좋아 보입니다 👏💯
몇 가지 코멘트 남겨두었습니다~ 확인부탁드려요!

### 고민한부분
1. BankClerk에 대해서는 코멘트에서 좀 더 자세하게 리뷰를 남겨드렸습니다. 
은행원이라면 신한은행을 다닐수도 있고, 우리은행을 다닐수도 있겠죠? 또 우리은행 용산지점을 다닐수도, 우리은행서울역 지점을 다닐수도 있을 것 같아요. 
따라서 저라면 은행원이 할 수 있는 것을 추상화시키고, 해당 은행원들은 어떠한 은행도 다닐 수 있도록 주입하는 방향도 고려해볼 것 같습니다. 🙏🏻

2. 해당 타입에서만 사용하는 값이라면 해당 타입이 그 값을 들고 있는 것이 좀 더 좋아보입니다. 만약 앱 내부에서 전반적으로 사용되는 공통적인 값이라면 말씀하신 것 처럼 해당 값들은 따로 분리하면 관리가 더 편할 것 같아요. 🙏🏻

3. BankManager관련해서도 코드에서 코멘트 남겨드렸습니다

ueunli committed:
chore: BankManager → ConsoleManager로 네이밍 변경

longlivedrgn committed:
chore: 가독성을 위한 메소드명 수정

- 기존의 closingMessage 메소드의 메소드 명을 printingClosingMessage으로 변경 및 argument label 추가

ueunli committed:
refactor: 폴더링

longlivedrgn committed:
chore: 메소드 네이밍 수정

- distributeCustomersToClerk에서 extractCustomerFromQueue로 함수명 변경
- BankClerk의 serve 메소드에 wildcard 활용

longlivedrgn committed:
Merge branch 'main' of https://github.com/ueunli/ios-bank-manager

ueunli committed:
refactor: Bank의 이니셜라이저에 매개변수 추가

- 호출부(외부)에서 은행원을 주입할 수 있도록 `[BankClerk]`타입의 매개변수 추가

longlivedrgn committed:
refactor: 무작위 손님 인원을 생성하는 기능을 Bank가 직접 수행하도록 변경

ueunli committed:
refactor: 은행폐점 메시지를 출력하는 메서드의 구현부 수정

- 네이밍: `printingClosingMessage(about:with:)` → `printClosingMessage(about:with:)`
- ConsoleManager의 static메서드 → Bank의 private메서드

jryoun1 commented on BankManagerConsoleApp/BankManagerConsoleApp/Bank/Bank.swift:
```diff
@@ -0,0 +1,51 @@
+    init(clerks: [BankClerk]) {
+        self.queue = Queue<String>()
+        self.clerks = clerks
+        self.numberOfCustomers = Int.random(in: rangeOfNumberOfCustomers.minimum...rangeOfNumberOfCustomers.maximum)
```
물론 말씀하신 것처럼 해당 은행에서 최소 몇명, 최대 몇명 사이의 랜덤의 숫자를 결정할 수도 있지만,
이 역시도 외부에서 주입을 받을 수 있으면 확장성이 고려될 것 같습니다. 

지금의 경우에는 이 Bank 개체는 무조건 10,30사이의 랜덤숫자를 생성하지만 만약 주입을 받는다면 실제 사용되는 곳에 따라서 혹은 runtime에 사용자의 입력에 따라서도 최소와 최대의 범위를 지정할 수 있을 것 같습니다.

jryoun1 reviewed (approved):
넵 코멘트 드린 내용들을 전반적으로 잘 수정해주신 것 같습니다. 
몇 가지 추가적인 부분들에 대해서는 기존 코멘트에 답글과 수정하신 부분에 코멘트를 남겼으니 다음 step에서 한 번 고민해보시면 좋을 것 같습니다. 

고생하셨습니다~ 이번 스텝은 머지하겠습니다!
다음 스텝도 화이팅입니다.
